optimize-onnx/src/optimize.cpp

```diff
@@ -0,0 +1,176 @@
+#include <queue>
+
+#include "optimize.h"
+
+namespace onnx { namespace optimization {
+
+bool isPureOperator(Node * n) {
+  // TODO This is a lie.
+  return true;
+}
+
+// Split the graph into 'init' and 'predict' nets. This is kind of
+// like constant folding, except that rather than actually execute the
+// constant computations, we simply split them out into a separate
+// graph. Nodes that have any transitive dependency on the
+// initializers, or on impure operators, must remain in the predict
+// net. All others may be moved to the init net.
+//
+// This function destructively mutates the graph into either the init
+// or the predict net. If you want both, which you probably do,
+// arrange to call it twice.
+//
+// NOTE POTENTIAL BREAKAGE:
+//
+// The ONNX spec provides no guarantees about "staging", i.e. which
+// inputs change on every invocation vs which generally stay the same.
+// Here we make the assumption that inputs which have an initializer
+// value provided for them vary only between invocations of the init
+// net, and are constant across runs of the predict net.
+//
+std::shared_ptr<Graph> split_init_and_predict(std::shared_ptr<Graph> g, bool init, bool predict) {
+  // The first step is a graph traversal to identify what is reachable
+  // from the mutable intializers. Since we have a bipartite graph, we
+  // must track both Nodes and Values that we have reached (as either
+  // could be terminal objects in the graph), but it's sufficient to
+  // maintain a frontier consisting only of Values.
+  std::unordered_set<Value *> live_values;
+  std::unordered_set<Node *> live_nodes;
```

Comment:
If you want to only have one map, then you can probably define "live node" to be (any inputs are live or it is not pure). But having two maps is fine as well.